src/components/navigation/navigation.tsx

```diff
@@ -0,0 +1,214 @@
+  //
+  //  Private Methods
+  //
+  //--------------------------------------------------------------------------
+
+  private clickHandler = () => {
+    this.calciteNavActionSelect.emit();
+  };
+
+  private handleContentSlotChange = (event: Event): void => {
+    const element = event.target as Element;
+    const level = element.getAttribute("data-level");
+    if (level === LEVEL.secondary) {
+      this.secondarySlotHasElements = slotChangeHasAssignedElement(event);
+    } else if (level === LEVEL.tertiary) {
+      this.tertiarySlotHasElements = slotChangeHasAssignedElement(event);
+    } else {
+      this.primarySlotHasElements = slotChangeHasAssignedElement(event);
+    }
+  };
+
+  private handleUserSlotChange = (event: Event): void => {
+    this.userSlotHasElements = slotChangeHasAssignedElement(event);
+  };
+
+  private handleLogoSlotChange = (event: Event): void => {
+    this.logoSlotHasElements = slotChangeHasAssignedElement(event);
+  };
+
+  private handleMenuActionSlotChange = (event: Event): void => {
+    const hasMenuAction = slotChangeHasAssignedElement(event);
+    if (hasMenuAction) {
+      this.navAction = false;
+    }
+  };
+
+  private hasSlottedElements(level: Level): boolean {
+    if (level === LEVEL.primary) {
+      return (
+        this.navAction ||
+        this.userSlotHasElements ||
+        this.logoSlotHasElements ||
+        this.primarySlotHasElements
+      );
+    } else if (level === LEVEL.secondary) {
+      return this.secondarySlotHasElements;
+    } else {
+      return this.tertiarySlotHasElements;
+    }
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Render Methods
+  //
+  //--------------------------------------------------------------------------
+
+  renderMenuAction(): VNode {
+    return (
+      <slot name={SLOTS.navAction} onSlotchange={this.handleMenuActionSlotChange}>
+        {this.navAction && (
+          <calcite-action icon="hamburger" onClick={this.clickHandler} text={this.label} />
+        )}
+      </slot>
+    );
+  }
+
+  renderNavLevel(level: Level): VNode {
+    const hasElements = this.hasSlottedElements(level);
+    const isPrimaryLevel = level === LEVEL.primary;
+    return (
+      <div
+        class={{
+          [CSS.container]: true,
+          [`nav-${level}`]: true,
+          hide: !hasElements
+        }}
+      >
+        <slot name={SLOTS.progress} />
+        <div class={CSS.containerContent}>
+          {isPrimaryLevel && this.renderMenuAction()}
+          {isPrimaryLevel && <slot name={SLOTS.logo} onSlotchange={this.handleLogoSlotChange} />}
+          <slot
+            data-level={level}
+            name={
+              isPrimaryLevel
+                ? SLOTS.primaryContentStart
+                : level === LEVEL.secondary
+                ? SLOTS.secondaryContentStart
+                : SLOTS.tertiaryContentStart
+            }
+            onSlotchange={this.handleContentSlotChange}
```

Comment:
Question: rather than having specific slots for levels, would it be more future proof and HTML like if the content could just be nested? 

This is more inline with other components we have like combobox-item where you can nest one inside the other in order to have levels.

Currently, it is hard restricted to 3 levels and requires a user to really define that level. It could also lead to users adding a tertiary level without a primary level.

More future proof would be to just allow nesting of components.

Reply:
cc @macandcheese

Reply:
Do you mean nesting additional nav components inside the parent nav? The primary reason for having the levels is to enforce nav logo and nav user at the top, as well as the primary nav has different height. If we add more support for nested nav, we’d probably need to introduce a scale to achieve the various levels, and this could break platform consistency quickly. 

What would the markup look like for nested nav? It would simplify the theming and css variable story which would be great, but we might lose consistency of implementation. We don’t want logo / user to ever reside under the primary navigation.

Not opposed if that makes the markup simpler and doesn’t affect intended design.

Reply:
I think the way it is is fine as long as we are committed to 3 levels.

Reply:
Took a quick shot at something like you describe - it simplifies the component a bit and seems to work well IMO - https://github.com/Esri/calcite-components/tree/macandcheese/slotted-navs

This ultimately may be more flexible - we can still can enforce certain slotted components in the "primary" nav, it will just require a bit of doc.